x-pack/plugins/security_solution/public/common/components/popover_items/index.tsx

```diff
@@ -0,0 +1,115 @@
+  width: 100%;
+`;
+
+const PopoverWrapper = styled(EuiBadgeGroup)`
+  max-height: 200px;
+  max-width: 600px;
+  overflow: auto;
+  line-height: ${({ theme }) => theme.eui.euiLineHeight};
+`;
+
+/**
+ * Component to render list of items in popover, wicth configurabe number of display items by default
+ * @param items - array of items to render
+ * @param renderItem - render function that render item, arguments: item, index, items[]
+ * @param popoverTitle - title of popover
+ * @param popoverButtonTitle - title of popover button that triggers popover
+ * @param popoverButtonIcon - icon of popover button that triggers popover
+ * @param numberOfItemsToDisplay - number of items to render that are no in popover, defaults to 0
+ * @param dataTestPrefix - data-test-subj prefix to apply to elements
+ */
+const PopoverItemsComponent = <T extends unknown>({
+  items,
+  renderItem,
+  popoverTitle,
+  popoverButtonTitle,
+  popoverButtonIcon,
+  numberOfItemsToDisplay = 0,
+  dataTestPrefix = 'items',
+}: PopoverItemsProps<T>) => {
+  const [isExceptionOverflowPopoverOpen, setIsExceptionOverflowPopoverOpen] = useState(false);
+
+  const OverflowList = ({ items: itemsToRender }: OverflowListProps<T>) => (
+    <>{itemsToRender.map(renderItem)}</>
+  );
+
+  if (items.length <= numberOfItemsToDisplay) {
+    return (
+      <PopoverItemsWrapper data-test-subj={dataTestPrefix} alignItems="center" gutterSize="s">
+        <OverflowList items={items} />
+      </PopoverItemsWrapper>
+    );
+  }
+
+  return (
+    <PopoverItemsWrapper alignItems="center" gutterSize="s" data-test-subj={dataTestPrefix}>
+      <EuiFlexItem grow={1} className="eui-textTruncate">
+        <OverflowList items={items.slice(0, numberOfItemsToDisplay)} />
+      </EuiFlexItem>
+      <EuiPopover
+        ownFocus
+        data-test-subj={`${dataTestPrefix}DisplayPopover`}
+        button={
+          <EuiBadge
+            iconType={popoverButtonIcon}
+            color="hollow"
+            data-test-subj={`${dataTestPrefix}DisplayPopoverButton`}
+            onClick={() => setIsExceptionOverflowPopoverOpen(!isExceptionOverflowPopoverOpen)}
+            onClickAriaLabel={popoverButtonTitle}
+          >
+            {popoverButtonTitle}
+          </EuiBadge>
+        }
+        isOpen={isExceptionOverflowPopoverOpen}
+        closePopover={() => setIsExceptionOverflowPopoverOpen(!isExceptionOverflowPopoverOpen)}
+        repositionOnScroll
+      >
+        {popoverTitle ? (
+          <EuiPopoverTitle data-test-subj={`${dataTestPrefix}DisplayPopoverTitle`}>
+            {popoverTitle}
+          </EuiPopoverTitle>
+        ) : null}
+        <PopoverWrapper data-test-subj={`${dataTestPrefix}DisplayPopoverWrapper`}>
+          <OverflowList items={items.slice(numberOfItemsToDisplay)} />
+        </PopoverWrapper>
+      </EuiPopover>
+    </PopoverItemsWrapper>
+  );
+};
+
+export const PopoverItems = React.memo(PopoverItemsComponent);
```

Comment:
When wrapping a generic react component with `React.memo`, its type arguments get "dropped". To restore them, you should cast the memoized component back to its original type:

```ts
export const PopoverItems = React.memo(PopoverItemsComponent) as typeof PopoverItemsComponent;
```

That will help you avoid using castings like `as PopoverItemsProps<unknown>['renderItem']` elsewhere in the codebase.

Reply:
Thanks for the nice suggestion. 
Changed it

Reply:
TIL, thanks @xcrzx! [Here's a big ole thread](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/37087#issuecomment-849434815) discussing that issue in case anyone is interested in the details here, gist being:

> TS doesn't currently support the ability to "forward" free type parameters when performing operations with higher order functions, so the free type parameter is lost when you create it,